src/i18n.ts

```diff
@@ -0,0 +1,342 @@
+		filename = matched;
+		return PATH_SEPARATOR;
+	});
+	const locales = getSupportedLocales(locale, supported);
+	return locales.map(function (locale: string): string {
+		return `${parentDirectory}${locale}${filename}`;
+	});
+}
+
+/**
+ * @private
+ * Ensure a path follows the required format for loading locale-specific bundles.
+ *
+ * @param path
+ * The default bundle path to validate.
+ */
+function validatePath(path: string): void {
+	if (!VALID_PATH_PATTERN.test(path)) {
+		const message = 'Invalid i18n bundle path. Bundle maps must adhere to the format' +
+			' "{basePath}{separator}{bundleName}" so that locale bundles can be resolved.';
+		throw new Error(message);
+	}
+}
+
+/**
+ * Load locale-specific messages for the specified bundle and locale.
+ *
+ * @param bundle
+ * The default bundle that is used to determine where the locale-specific bundles are located.
+ *
+ * @param context
+ * An optional locale, or stateful object that specifies a string locale. If the context is a stateful object with an
+ * `invalidate` method, it will be invalidated when the locale is changed. If no context is provided, then the current
+ * locale is assumed.
+ *
+ * @return A promise to the locale-specific messages.
+ */
+function i18n<T extends Messages>(bundle: Bundle<T>): Promise<T>;
+function i18n<T extends Messages>(bundle: Bundle<T>, context: LocaleContext): Promise<T>;
+function i18n<T extends Messages>(bundle: Bundle<T>, locale: string): Promise<T>;
+function i18n<T extends Messages>(bundle: Bundle<T>, context?: any): Promise<T> {
+	const locale = resolveLocale(context);
+	const { baseUrl, locales, messages } = bundle;
+	const path = baseUrl.replace(/\/$/, '');
+
+	try {
+		validatePath(path);
+	}
+	catch (error) {
+		return Promise.reject(error);
+	}
+
+	if (context && isContextObject(context) && contextObjects.indexOf(context) === -1) {
+		contextObjects.push(context);
+		context.own(<Handle> {
+			destroy(this: Handle) {
+				this.destroy = () => {};
+				const index = contextObjects.indexOf(context);
+				contextObjects.splice(index, 1);
+			}
+		});
+	}
+
+	const localePaths = resolveLocalePaths(path, locale, locales);
+
+	if (!localePaths.length) {
+		return Promise.resolve(messages);
+	}
+
+	return loadLocaleBundles(localePaths).then(function (bundles: T[]): T {
+		return bundles.reduce(function (previous: T, partial: T, i: number): T {
+			const localeMessages = assign({}, previous, partial);
+			return localeMessages;
+		}, messages);
+	});
+}
+
+export default i18n;
+
+/**
+ * Change the root locale, and invalidate any registered statefuls.
+ *
+ * @param locale
+ * The new locale.
+ */
+export function switchLocale(locale: string): void {
+	const previous = rootLocale;
+	rootLocale = locale;
+
+	if (previous !== rootLocale) {
+		contextObjects.forEach(function (context: LocaleContext) {
+			if (!context.state.locale) {
+				(<any> context).invalidate();
```

Comment:
You should be able to deal with the strict null checks with `context.invalidate && context.invalidate();`